telemetryMetrics improvements

rclee opened the pull request:
This PR implements [schema changes](https://github.com/mapbox/event-schema/commit/74f60cdefcdb967ff16f8e61f71f06f90aec5584) for `telemetryMetrics` to begin tracking received data and fixes some bugs with event counting. 

Fixes https://github.com/mapbox/mobile-telemetry/issues/303

rclee committed:
Reformat failedRequests metric 

This should allow the failedRequests metric to be more queryable.

rclee committed:
Track received data 

Implement the event schema changes to begin tracking received data on all requests. 

Tested using breakpoints to track each `data.length` from start to finish.

rclee committed:
Fix bug in event counting

Although this wouldn't be much an issue in our current implementation; once we add in deferred telemetry we would have missed counting any additional batches in APIClient.

Also adds a failed event count if the request is ever nil.

alfwatt commented on MapboxMobileEvents/MMEMetricsManager.h:
```diff
@@ -10,9 +10,10 @@ NS_ASSUME_NONNULL_BEGIN
 
-- (void)updateMetricsFromData:(NSData *)data;
+- (void)updateTransferredData:(NSData *)data;
+- (void)updateReceivedData:(NSData *)data;
```
As above suggest `NSUInteger bytes` and `updateReceivedBytes`

alfwatt commented on MapboxMobileEvents/MMEMetricsManager.m:
```diff
@@ -49,37 +49,63 @@ - (void)updateMetricsFromEventQueue:(NSArray *)eventQueue {
+    }
+}
+
+- (void)updateEventsFailedCountFromEvents:(NSArray *)events {
```
Suggest `NSUInteger eventCount` and rename to `updateEventsFailedCount`

alfwatt commented on MapboxMobileEvents/MMEMetrics.h:
```diff
@@ -9,6 +9,9 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic) NSUInteger totalDataTransfer;
 @property (nonatomic) NSUInteger cellDataTransfer;
 @property (nonatomic) NSUInteger wifiDataTransfer;
```
Suggest making `readonly`, renaming to `wifiBytesTransfered` and computing the value

alfwatt reviewed (commented):
Suggestions inline, generally: try to be DRY about your data storage, computing the totals makes it easier to keep the test passing and saves a little memory at the expense of a little runtime.

Better to pass in the values, instead of the containers (NSData/NSArray) to prevent accidental retention. Method/property names should reflect the units they're recording (Bytes vs. Data for e.g.)

rclee committed:
Implement suggested changes 

Also adds missing tests and removed the synchronous NSURLConnection call that was being used for testing. 

Changed transferred bytes to sent bytes -- this will be reflected in the schema soon.

rclee committed:
Move telemetryMetrics send request method

Moves `telemetryMetrics` send request method to `flush` to have the event send more regularly. 

Previously this method was moved out of `flush` to be more readable but we missed a vital part that `flush` is also called on a timer and `sendTelemetryMetricsEvent` was out of that loop.

alfwatt reviewed (commented):
Sorry I was a little unclear about the `readonly` properties, `computed` might be a better way to describe them than `synthetic` which means something else in Java for e.g.: when the property getter is called, you can do the math real quick, instead of looking at the stored value.

Slightly more compute time but less memory use, and no chance of the values getting out of synch.

rclee committed:
Add better bytes compute getter

alfwatt reviewed (dismissed):
Looks good, just needs the test cases around the timer and we're good to commit!

alfwatt reviewed (commented):
Also, you should rebase this branch from master:

`git fetch`
`git rebase master`

There are a couple of small conflicts where I'd renamed `date` to `recordingStarted` that you'll need to resole by hand. if you run into trouble (it can be easy to end up on a detached branch if you don't follow the rebase procedure just so) let me know and we can pair up

rclee committed:
Merge branch 'master' into rclee-telemetrymetrics-improvements

rclee committed:
Rename date to recordingStarted

rclee committed:
Add telemetryMetrics timer test

Adds a unit test to ensure that we attempt to send a telemetryMetrics event when the event queue timer fires.
